lib/balanced-pool.js

```diff
@@ -62,13 +103,48 @@ class BalancedPool extends Dispatcher {
       throw new BalancedPoolMissingUpstreamError()
     }
 
-    let pool = this[kPools].shift()
-    this[kPools].push(pool)
-    if (pool.busy) {
-      pool = this[kPools].find(pool => !pool.busy) || pool
+    const dispatchPool = (pool) => {
+      this[kNeedDrain] = !pool.dispatch(opts, handler)
+      return !this[kNeedDrain]
+    }
+
+    const allPoolsBusy = this[kPools].reduce((a, b) => a && b, true)
+    let nextWeightIndex
+    let counter = 0
+
+    while (counter++ < this[kPools].length) {
+      this[kIndex] = (this[kIndex] + 1) % this[kPools].length
+      const pool = this[kPools][this[kIndex]]
+
+      // Save the index of the pool with the next lower weight
+      // This is needed incase we couldn't find a pool that is >=currentWeight in `this[kpools].length` iterations
+      // This is an optimization. it also avoids using an infinite loop as the algorithim described in the paper.
+      if (pool[kWeight] < this[kCurrentWeight] && (!pool.busy || allPoolsBusy)) {
```

Comment:
@ronag @mcollina  I had to use add this logic to avoid using a while(true)

What do you think?

Reply:
pool.busy is not a defined property